### ISSUE TITLE
Use of container_of()

### ISSUE THREAD

**Issue by yzgyyang** [iovisor/bcc#3314]
Hi @yonghong-song, I saw that you committed https://www.spinics.net/lists/bpf/msg18493.html a while ago which hinted that `offsetof()` and `container_of()` should work in BPF. I am trying a similar thing with them and cannot get it to work. Not sure if this is the right place to ask, I apologize in advance if it's not. Also, the following is just my understanding, could be wrong :)

### Background

#237 - in order to get the file path, we need a `path->dentry->d_parent` walk to the root. However, fully-qualified paths cannot be derived if the dentry is on a mounted filesystem (like NFS) - the walk will stop before the mountpoint (e.g. if `/mnt/nfs` is an NFS mount, then `/mnt/nfs/a/b/c.txt` will only reveal to `/a/b/c.txt` for the walk). So the goal is to get the mountpoint name to complement the partial path that obtained.

`vfs_*()` calls have an argument `struct path *p`. Some definitions here:

```c
struct path {
	struct vfsmount *mnt;
	struct dentry *dentry;
} __randomize_layout;

struct vfsmount {
	struct dentry *mnt_root;	/* root of the mounted tree */
	struct super_block *mnt_sb;	/* pointer to superblock */
	int mnt_flags;
	struct user_namespace *mnt_userns;
} __randomize_layout;
```

So the idea (as suggested in https://github.com/torvalds/linux/blob/v5.8/fs/namei.c#L1102) is to use `container_of()` macro to get the `struct mount` that contained `struct vfsmount` above, then we will be able to get `mount->mnt_mountpoint` which is the dentry for the mountpoint. (`vfsmount->mnt_root->d_name.name == "/"`)

```c
struct mount {
	struct hlist_node mnt_hash;
	struct mount *mnt_parent;
	struct dentry *mnt_mountpoint;
	struct vfsmount mnt;
        ...
} __randomize_layout;
```

### Problem

The goal is to get the mountpoint name to complement the partial path that obtained by `container_of()`. I have the following BPF code:

```c
#include <uapi/linux/ptrace.h>
#include <linux/blkdev.h>
#include <linux/dcache.h>
#include <linux/mount.h>

struct mount {
	struct hlist_node mnt_hash;
	struct mount *mnt_parent;
	struct dentry *mnt_mountpoint;
	struct vfsmount mnt;
};

struct mountpoint {
	struct hlist_node m_hash;
	struct dentry *m_dentry;
	struct hlist_head m_list;
	int m_count;
};

#ifndef offsetof
#define offsetof(TYPE, MEMBER)  ((size_t)&((TYPE *)0)->MEMBER)
#endif

#ifndef containerof
#define containerof(ptr, type, member) ({                      \
    const typeof( ((type *)0)->member ) *__mptr = (ptr);    \
    (type *)( (char *)__mptr - offsetof(type,member) );})
#endif

int trace_entry(struct pt_regs *ctx, const struct path *p,
    struct kstat *k, u32 d, unsigned int c)
{
    ...
    // find mountpoint
    struct vfsmount *vmnt;
    bpf_probe_read_kernel(&vmnt, sizeof(vmnt), p->mnt);
    struct mount *real_mnt = containerof(vmnt, struct mount, mnt);
    struct mountpoint *m_mp = real_mnt->mnt_mp;
    struct dentry *m_de = m_mp->m_dentry;
    struct qstr m_dname = m_de->d_name;
    info.mountpoint_len = m_dname.len; // XXX: this fails
    bpf_probe_read_kernel(&info.mountpoint, sizeof(info.mountpoint), m_dname.name); // XXX: this fails
    ...
    return 0;
}
```

With:

```python
b.attach_kprobe(event="vfs_open", fn_name="trace_entry")
```

But then I got:
```
bpf: Failed to load program: Permission denied
0: (79) r7 = *(u64 *)(r1 +112)
1: (bf) r1 = r10
2: (07) r1 += -16
3: (b7) r2 = 16
4: (85) call bpf_get_current_comm#16
5: (71) r1 = *(u8 *)(r10 -16)
6: (55) if r1 != 0x6e goto pc+349
 R0=inv(id=0) R1=inv110 R7=inv(id=0) R10=fp0,call_-1
7: (71) r1 = *(u8 *)(r10 -15)
8: (15) if r1 == 0x66 goto pc+1
 R0=inv(id=0) R1=inv(id=0,umax_value=255,var_off=(0x0; 0xff)) R7=inv(id=0) R10=fp0,call_-1
9: (05) goto pc+346
356: (b7) r0 = 0
357: (95) exit

from 8 to 10: R0=inv(id=0) R1=inv102 R7=inv(id=0) R10=fp0,call_-1
10: (71) r1 = *(u8 *)(r10 -14)
11: (55) if r1 != 0x73 goto pc+344
 R0=inv(id=0) R1=inv115 R7=inv(id=0) R10=fp0,call_-1
12: (71) r1 = *(u8 *)(r10 -13)
13: (55) if r1 != 0x64 goto pc+342
 R0=inv(id=0) R1=inv100 R7=inv(id=0) R10=fp0,call_-1
14: (71) r1 = *(u8 *)(r10 -12)
15: (55) if r1 != 0x0 goto pc+340
 R0=inv(id=0) R1=inv0 R7=inv(id=0) R10=fp0,call_-1
16: (71) r1 = *(u8 *)(r10 -11)
17: (15) if r1 == 0x0 goto pc+1
 R0=inv(id=0) R1=inv(id=0,umax_value=255,var_off=(0x0; 0xff)) R7=inv(id=0) R10=fp0,call_-1
18: (05) goto pc+337
356: safe

from 17 to 19: R0=inv(id=0) R1=inv0 R7=inv(id=0) R10=fp0,call_-1
19: (b7) r6 = 0
20: (7b) *(u64 *)(r10 -432) = r6
21: (bf) r8 = r7
22: (07) r8 += 8
23: (bf) r1 = r10
24: (07) r1 += -432
25: (b7) r2 = 8
26: (bf) r3 = r8
27: (85) call bpf_probe_read#4
28: (79) r3 = *(u64 *)(r10 -432)
29: (7b) *(u64 *)(r10 -424) = r6
30: (7b) *(u64 *)(r10 -432) = r6
31: (07) r3 += 32
32: (bf) r1 = r10
33: (07) r1 += -432
34: (b7) r2 = 16
35: (85) call bpf_probe_read#4
36: (61) r9 = *(u32 *)(r10 -428)
37: (15) if r9 == 0x0 goto pc+318
 R0=inv(id=0) R6=inv0 R7=inv(id=0) R8=inv(id=0) R9=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R10=fp0,call_-1
38: (79) r1 = *(u64 *)(r10 -424)
39: (7b) *(u64 *)(r10 -472) = r1
40: (63) *(u32 *)(r10 -24) = r6
41: (7b) *(u64 *)(r10 -32) = r6
42: (7b) *(u64 *)(r10 -40) = r6
43: (7b) *(u64 *)(r10 -48) = r6
44: (7b) *(u64 *)(r10 -56) = r6
45: (7b) *(u64 *)(r10 -64) = r6
46: (7b) *(u64 *)(r10 -72) = r6
47: (7b) *(u64 *)(r10 -80) = r6
48: (7b) *(u64 *)(r10 -88) = r6
49: (7b) *(u64 *)(r10 -96) = r6
50: (7b) *(u64 *)(r10 -104) = r6
51: (7b) *(u64 *)(r10 -112) = r6
52: (7b) *(u64 *)(r10 -120) = r6
53: (7b) *(u64 *)(r10 -128) = r6
54: (7b) *(u64 *)(r10 -136) = r6
55: (7b) *(u64 *)(r10 -144) = r6
56: (7b) *(u64 *)(r10 -152) = r6
57: (7b) *(u64 *)(r10 -160) = r6
58: (7b) *(u64 *)(r10 -168) = r6
59: (7b) *(u64 *)(r10 -176) = r6
60: (7b) *(u64 *)(r10 -184) = r6
61: (7b) *(u64 *)(r10 -192) = r6
62: (7b) *(u64 *)(r10 -200) = r6
63: (7b) *(u64 *)(r10 -208) = r6
64: (7b) *(u64 *)(r10 -216) = r6
65: (7b) *(u64 *)(r10 -224) = r6
66: (7b) *(u64 *)(r10 -232) = r6
67: (7b) *(u64 *)(r10 -240) = r6
68: (7b) *(u64 *)(r10 -248) = r6
69: (7b) *(u64 *)(r10 -256) = r6
70: (7b) *(u64 *)(r10 -264) = r6
71: (7b) *(u64 *)(r10 -272) = r6
72: (7b) *(u64 *)(r10 -280) = r6
73: (7b) *(u64 *)(r10 -288) = r6
74: (7b) *(u64 *)(r10 -296) = r6
75: (7b) *(u64 *)(r10 -304) = r6
76: (7b) *(u64 *)(r10 -312) = r6
77: (7b) *(u64 *)(r10 -320) = r6
78: (7b) *(u64 *)(r10 -328) = r6
79: (7b) *(u64 *)(r10 -336) = r6
80: (7b) *(u64 *)(r10 -344) = r6
81: (7b) *(u64 *)(r10 -352) = r6
82: (7b) *(u64 *)(r10 -360) = r6
83: (7b) *(u64 *)(r10 -368) = r6
84: (7b) *(u64 *)(r10 -376) = r6
85: (7b) *(u64 *)(r10 -384) = r6
86: (7b) *(u64 *)(r10 -392) = r6
87: (7b) *(u64 *)(r10 -400) = r6
88: (7b) *(u64 *)(r10 -408) = r6
89: (7b) *(u64 *)(r10 -416) = r6
90: (7b) *(u64 *)(r10 -424) = r6
91: (7b) *(u64 *)(r10 -432) = r6
92: (71) r1 = *(u8 *)(r10 -16)
93: (73) *(u8 *)(r10 -432) = r1
94: (bf) r1 = r10
95: (07) r1 += -432
96: (7b) *(u64 *)(r10 -464) = r1
97: (b7) r2 = 16
98: (85) call bpf_get_current_comm#16
99: (63) *(u32 *)(r10 -384) = r9
100: (bf) r1 = r10
101: (07) r1 += -416
102: (b7) r2 = 32
103: (79) r3 = *(u64 *)(r10 -472)
104: (85) call bpf_probe_read#4
105: (7b) *(u64 *)(r10 -456) = r6
106: (bf) r1 = r10
107: (07) r1 += -456
108: (b7) r2 = 8
109: (bf) r3 = r7
110: (85) call bpf_probe_read#4
111: (79) r3 = *(u64 *)(r10 -456)
112: (bf) r1 = r10
113: (07) r1 += -440
114: (b7) r2 = 8
115: (85) call bpf_probe_read#4
116: (79) r1 = *(u64 *)(r10 -440)
117: (79) r1 = *(u64 *)(r1 +192)
R1 invalid mem access 'inv'

HINT: The invalid mem access 'inv' error can happen if you try to dereference memory without first using bpf_probe_read_kernel() to copy it to the BPF stack. Sometimes the bpf_probe_read_kernel() is automatic by the bcc rewriter, other times you'll need to be explicit.
```

Do you have any insights/suggestions on this? Thanks!

**Comment by yonghong-song** [comment]
container_of should work in this case. For the above verifier failure, I think,
       struct mountpoint *m_mp = real_mnt->mnt_mp;
need explicit bpf_probe_read_kernel. Maybe the rest needs too. The rewriter lost track of pointer chasing so not able to rewriter this with proper probe helpers.

**Comment by yzgyyang** [comment]
@yonghong-song thanks for the reply! I tried rewriting those to `bpf_probe_read_kernel()`, and the verifier now throws errors at the `struct mount` level (the "parent" container we got from `containerof()`) instead of the bottom of pointer chasing:

```c
// find mountpoint
struct vfsmount *vmnt;
bpf_probe_read_kernel(&vmnt, sizeof(vmnt), p->mnt);
struct mount *real_mnt;
bpf_probe_read_kernel(&real_mnt, sizeof(real_mnt), containerof(vmnt, struct mount, mnt));
struct mountpoint *m_mp;
bpf_probe_read_kernel(&m_mp, sizeof(m_mp), real_mnt->mnt_mp); // XXX: this fails
struct dentry *m_de;
bpf_probe_read_kernel(&m_de, sizeof(m_de), real_mnt->mnt_mountpoint); // XXX: this fails
```

The error looks like:
```
bpf: Failed to load program: Permission denied
0: (79) r7 = *(u64 *)(r1 +112)
1: (bf) r1 = r10
2: (07) r1 += -16
3: (b7) r2 = 16
4: (85) call bpf_get_current_comm#16
5: (71) r1 = *(u8 *)(r10 -16)
6: (55) if r1 != 0x6e goto pc+356
 R0=inv(id=0) R1=inv110 R7=inv(id=0) R10=fp0,call_-1
7: (71) r1 = *(u8 *)(r10 -15)
8: (15) if r1 == 0x66 goto pc+1
 R0=inv(id=0) R1=inv(id=0,umax_value=255,var_off=(0x0; 0xff)) R7=inv(id=0) R10=fp0,call_-1
9: (05) goto pc+353
363: (b7) r0 = 0
364: (95) exit

from 8 to 10: R0=inv(id=0) R1=inv102 R7=inv(id=0) R10=fp0,call_-1
10: (71) r1 = *(u8 *)(r10 -14)
11: (55) if r1 != 0x73 goto pc+351
 R0=inv(id=0) R1=inv115 R7=inv(id=0) R10=fp0,call_-1
12: (71) r1 = *(u8 *)(r10 -13)
13: (55) if r1 != 0x64 goto pc+349
 R0=inv(id=0) R1=inv100 R7=inv(id=0) R10=fp0,call_-1
14: (71) r1 = *(u8 *)(r10 -12)
15: (55) if r1 != 0x0 goto pc+347
 R0=inv(id=0) R1=inv0 R7=inv(id=0) R10=fp0,call_-1
16: (71) r1 = *(u8 *)(r10 -11)
17: (15) if r1 == 0x0 goto pc+1
 R0=inv(id=0) R1=inv(id=0,umax_value=255,var_off=(0x0; 0xff)) R7=inv(id=0) R10=fp0,call_-1
18: (05) goto pc+344
363: safe

from 17 to 19: R0=inv(id=0) R1=inv0 R7=inv(id=0) R10=fp0,call_-1
19: (b7) r6 = 0
20: (7b) *(u64 *)(r10 -432) = r6
21: (bf) r8 = r7
22: (07) r8 += 8
23: (bf) r1 = r10
24: (07) r1 += -432
25: (b7) r2 = 8
26: (bf) r3 = r8
27: (85) call bpf_probe_read#4
28: (79) r3 = *(u64 *)(r10 -432)
29: (7b) *(u64 *)(r10 -424) = r6
30: (7b) *(u64 *)(r10 -432) = r6
31: (07) r3 += 32
32: (bf) r1 = r10
33: (07) r1 += -432
34: (b7) r2 = 16
35: (85) call bpf_probe_read#4
36: (61) r9 = *(u32 *)(r10 -428)
37: (15) if r9 == 0x0 goto pc+325
 R0=inv(id=0) R6=inv0 R7=inv(id=0) R8=inv(id=0) R9=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R10=fp0,call_-1
38: (79) r1 = *(u64 *)(r10 -424)
39: (7b) *(u64 *)(r10 -488) = r1
40: (63) *(u32 *)(r10 -24) = r6
41: (7b) *(u64 *)(r10 -32) = r6
42: (7b) *(u64 *)(r10 -40) = r6
43: (7b) *(u64 *)(r10 -48) = r6
44: (7b) *(u64 *)(r10 -56) = r6
45: (7b) *(u64 *)(r10 -64) = r6
46: (7b) *(u64 *)(r10 -72) = r6
47: (7b) *(u64 *)(r10 -80) = r6
48: (7b) *(u64 *)(r10 -88) = r6
49: (7b) *(u64 *)(r10 -96) = r6
50: (7b) *(u64 *)(r10 -104) = r6
51: (7b) *(u64 *)(r10 -112) = r6
52: (7b) *(u64 *)(r10 -120) = r6
53: (7b) *(u64 *)(r10 -128) = r6
54: (7b) *(u64 *)(r10 -136) = r6
55: (7b) *(u64 *)(r10 -144) = r6
56: (7b) *(u64 *)(r10 -152) = r6
57: (7b) *(u64 *)(r10 -160) = r6
58: (7b) *(u64 *)(r10 -168) = r6
59: (7b) *(u64 *)(r10 -176) = r6
60: (7b) *(u64 *)(r10 -184) = r6
61: (7b) *(u64 *)(r10 -192) = r6
62: (7b) *(u64 *)(r10 -200) = r6
63: (7b) *(u64 *)(r10 -208) = r6
64: (7b) *(u64 *)(r10 -216) = r6
65: (7b) *(u64 *)(r10 -224) = r6
66: (7b) *(u64 *)(r10 -232) = r6
67: (7b) *(u64 *)(r10 -240) = r6
68: (7b) *(u64 *)(r10 -248) = r6
69: (7b) *(u64 *)(r10 -256) = r6
70: (7b) *(u64 *)(r10 -264) = r6
71: (7b) *(u64 *)(r10 -272) = r6
72: (7b) *(u64 *)(r10 -280) = r6
73: (7b) *(u64 *)(r10 -288) = r6
74: (7b) *(u64 *)(r10 -296) = r6
75: (7b) *(u64 *)(r10 -304) = r6
76: (7b) *(u64 *)(r10 -312) = r6
77: (7b) *(u64 *)(r10 -320) = r6
78: (7b) *(u64 *)(r10 -328) = r6
79: (7b) *(u64 *)(r10 -336) = r6
80: (7b) *(u64 *)(r10 -344) = r6
81: (7b) *(u64 *)(r10 -352) = r6
82: (7b) *(u64 *)(r10 -360) = r6
83: (7b) *(u64 *)(r10 -368) = r6
84: (7b) *(u64 *)(r10 -376) = r6
85: (7b) *(u64 *)(r10 -384) = r6
86: (7b) *(u64 *)(r10 -392) = r6
87: (7b) *(u64 *)(r10 -400) = r6
88: (7b) *(u64 *)(r10 -408) = r6
89: (7b) *(u64 *)(r10 -416) = r6
90: (7b) *(u64 *)(r10 -424) = r6
91: (7b) *(u64 *)(r10 -432) = r6
92: (71) r1 = *(u8 *)(r10 -16)
93: (73) *(u8 *)(r10 -432) = r1
94: (bf) r1 = r10
95: (07) r1 += -432
96: (7b) *(u64 *)(r10 -480) = r1
97: (b7) r2 = 16
98: (85) call bpf_get_current_comm#16
99: (63) *(u32 *)(r10 -384) = r9
100: (bf) r1 = r10
101: (07) r1 += -416
102: (b7) r2 = 32
103: (79) r3 = *(u64 *)(r10 -488)
104: (85) call bpf_probe_read#4
105: (7b) *(u64 *)(r10 -472) = r6
106: (bf) r1 = r10
107: (07) r1 += -472
108: (b7) r2 = 8
109: (bf) r3 = r7
110: (85) call bpf_probe_read#4
111: (79) r3 = *(u64 *)(r10 -472)
112: (bf) r1 = r10
113: (07) r1 += -440
114: (b7) r2 = 8
115: (85) call bpf_probe_read#4
116: (79) r3 = *(u64 *)(r10 -440)
117: (07) r3 += -32
118: (bf) r1 = r10
119: (07) r1 += -448
120: (b7) r2 = 8
121: (85) call bpf_probe_read#4
122: (79) r1 = *(u64 *)(r10 -448)
123: (79) r3 = *(u64 *)(r1 +24)
R1 invalid mem access 'inv'

HINT: The invalid mem access 'inv' error can happen if you try to dereference memory without first using bpf_probe_read_kernel() to copy it to the BPF stack. Sometimes the bpf_probe_read_kernel() is automatic by the bcc rewriter, other times you'll need to be explicit.
```

**Comment by yonghong-song** [comment]
Could you use explicit probe read helper for real_mnt->mnt_mp and real_mnt->mnt_mountpoint? After using container_of, rewriter probably losts track of pointer tracing and not able to do proper rewrite any more.

```
struct mountpoint *m_mp;
bpf_probe_read_kernel(&m_mp, sizeof(m_mp), real_mnt->mnt_mp); // XXX: this fails
struct dentry *m_de;
bpf_probe_read_kernel(&m_de, sizeof(m_de), real_mnt->mnt_mountpoint); // XXX: this fails
```

**Comment by yzgyyang** [comment]
> Could you use explicit probe read helper for real_mnt->mnt_mp and real_mnt->mnt_mountpoint?

@yonghong-song Sorry, I didn't follow - which `explicit probe read helper` are you referring to? I tried `bpf_probe_read_kernel()` and `bpf_probe_read()` without any luck. :(

**Comment by yonghong-song** [comment]
I mean adding something like below
```
bpf_probe_read_kernel(&mnt_mp, sizeof(mnt_mp), &real_mnt->mnt_mp);
bpf_probe_read_kernel(&m_mp, sizeof(m_mp), mnt_mp);
...
```

**Comment by yzgyyang** [comment]
@yonghong-song Thanks - I added that for `mnt_mp`, still the same error :(

```c
    // find mountpoint
    struct vfsmount *vmnt;
    bpf_probe_read_kernel(&vmnt, sizeof(vmnt), p->mnt);
    struct mount *real_mnt;
    bpf_probe_read_kernel(&real_mnt, sizeof(real_mnt), containerof(vmnt, struct mount, mnt));

    struct dentry *mnt_de;
    bpf_probe_read_kernel(&mnt_de, sizeof(mnt_de), &real_mnt->mnt_mountpoint);
    struct dentry *m_de;
    bpf_probe_read_kernel(&m_de, sizeof(m_de), mnt_de);
    struct qstr m_dname = m_de->d_name;
    bpf_probe_read_kernel(&info.mountpoint, sizeof(info.mountpoint), m_dname.name); // XXX: This fails
```

For testing, here is the full code:

```c
#include <uapi/linux/ptrace.h>
#include <linux/blkdev.h>
#include <linux/dcache.h>
#include <linux/mount.h>

struct mount {
	struct hlist_node mnt_hash;
	struct mount *mnt_parent;
	struct dentry *mnt_mountpoint;
	struct vfsmount mnt;
	struct mountpoint *mnt_mp;
};

struct mountpoint {
	struct hlist_node m_hash;
	struct dentry *m_dentry;
	struct hlist_head m_list;
	int m_count;
};

/*
 * Helper macro to manipulate data structures
 */
#ifndef offsetof
#define offsetof(TYPE, MEMBER)  ((size_t)&((TYPE *)0)->MEMBER)
#endif

#ifndef containerof
#define containerof(ptr, type, member) ({                      \
    const typeof( ((type *)0)->member ) *__mptr = (ptr);    \
    (type *)( (char *)__mptr - offsetof(type,member) );})
#endif

// the key for the output summary
struct info_t {
    char mountpoint[DNAME_INLINE_LEN];
};

// the value of the output summary
struct val_t {
    u64 opens;
};

BPF_HASH(counts, struct info_t, struct val_t);

int trace_open_entry(struct pt_regs *ctx, const struct path *p,
    struct kstat *k, u32 d, unsigned int c)
{
    // skip I/O lacking a filename
    struct dentry *de = p->dentry;
    struct qstr d_name = de->d_name;
    if (d_name.len == 0)
        return 0;

    // store counts
    struct info_t info;

    // find mountpoint
    struct vfsmount *vmnt;
    bpf_probe_read_kernel(&vmnt, sizeof(vmnt), p->mnt);
    struct mount *real_mnt;
    bpf_probe_read_kernel(&real_mnt, sizeof(real_mnt), containerof(vmnt, struct mount, mnt));

    struct dentry *mnt_de;
    bpf_probe_read_kernel(&mnt_de, sizeof(mnt_de), &real_mnt->mnt_mountpoint);
    struct dentry *m_de;
    bpf_probe_read_kernel(&m_de, sizeof(m_de), mnt_de);
    struct qstr m_dname = m_de->d_name;
    bpf_probe_read_kernel(&info.mountpoint, sizeof(info.mountpoint), m_dname.name); // XXX: This fails

    struct val_t *valp, zero = {};
    valp = counts.lookup_or_try_init(&info, &zero);
    if (valp) {
        valp->opens++;
    }
    return 0;
}
```

**Comment by yzgyyang** [comment]
@yonghong-song Thanks! The code now passes the verifier. I still cannot extract any mountpoint-related info, and actually found another issue:

Originally, I partially redefined `struct mount` since those structures in `fs/mount.h` are not exported. This is similar to what [mountsnoop.py](https://github.com/iovisor/bcc/blob/7e3f0c08c7c28757711c0a173b5bd7d9a31cf7ee/tools/mountsnoop.py) does:

```c
/*
 * XXX: struct mnt_namespace is defined in fs/mount.h, which is private to the
 * VFS and not installed in any kernel-devel packages. So, let's duplicate the
 * important part of the definition. There are actually more members in the
 * real struct, but we don't need them, and they're more likely to change.
 */
struct mnt_namespace {
    atomic_t count;
    struct ns_common ns;
};
```

`struct mount` in `fs/mount.h` looks like:

```c
struct mount {
	struct hlist_node mnt_hash;
	struct mount *mnt_parent;
	struct dentry *mnt_mountpoint;
	struct vfsmount mnt;
	union {
		struct rcu_head mnt_rcu;
		struct llist_node mnt_llist;
	};
#ifdef CONFIG_SMP
	struct mnt_pcp __percpu *mnt_pcp;
#else
	int mnt_count;
	int mnt_writers;
#endif
	struct list_head mnt_mounts;	/* list of children, anchored here */
	struct list_head mnt_child;	/* and going through their mnt_child */
	struct list_head mnt_instance;	/* mount instance on sb->s_mounts */
	const char *mnt_devname;	/* Name of device e.g. /dev/dsk/hda1 */
	struct list_head mnt_list;
	struct list_head mnt_expire;	/* link in fs-specific expiry list */
	struct list_head mnt_share;	/* circular list of shared mounts */
	struct list_head mnt_slave_list;/* list of slave mounts */
	struct list_head mnt_slave;	/* slave list entry */
	struct mount *mnt_master;	/* slave is on master->mnt_slave_list */
	struct mnt_namespace *mnt_ns;	/* containing namespace */
	struct mountpoint *mnt_mp;	/* where is it mounted */
	struct hlist_node mnt_mp_list;	/* list mounts with the same mountpoint */
	struct list_head mnt_umounting; /* list entry for umount propagation */
#ifdef CONFIG_FSNOTIFY
	struct fsnotify_mark_connector __rcu *mnt_fsnotify_marks;
	__u32 mnt_fsnotify_mask;
#endif
	int mnt_id;			/* mount identifier */
	int mnt_group_id;		/* peer group identifier */
	int mnt_expiry_mark;		/* true if marked for expiry */
	struct hlist_head mnt_pins;
	struct fs_pin mnt_umount;
	struct dentry *mnt_ex_mountpoint;
} __randomize_layout;
```

So that I have a partial redefinition in the BPF C code with only the parts that I might need:

```c
struct mount {
	struct hlist_node mnt_hash;
	struct mount *mnt_parent;
	struct dentry *mnt_mountpoint;
	struct vfsmount mnt;
	int mnt_id;			/* mount identifier */
};
```

In attempts to retrieve info for the mountpoint, first I tried `real_mnt->mnt_id` (output in the `mnt_id` field):

```c
    // find mountpoint
    struct vfsmount *vmnt;
    bpf_probe_read_kernel(&vmnt, sizeof(vmnt), p->mnt);
    struct mount *real_mnt;
    bpf_probe_read_kernel(&real_mnt, sizeof(real_mnt), containerof(vmnt, struct mount, mnt));
    bpf_probe_read_kernel(&info.mnt_id, sizeof(info.mnt_id), &real_mnt->mnt_id);
```

The below output (`mnt_id` field) shows that at least **some** info for the mount is passed back in `mnt_id` (files on the same mount has the same value), it is just nowhere near the actual `mnt_id` values. Additionally, those values change when I change the partial `struct mount` definition. This leads me to suspect that the offset for `real_mnt->mnt_id` is not calculated correctly and the wrong value was retrieved.

```
COMM          OPEN    MOUNTPOINT         PATH                                   mnt_id
nfsd            3          b''        /fixtures                               938580992
nfsd            1          b''        /fixtures/foo.txt                        938580992
nfsd            1          b''        /                                      938580992
nfsd            1          b''        /fixtures/about_us.md                    938580992
nfsd            1          b''        /fixtures/loremipsum.txt                 938580992
nfsd            1          b''        /vol/share                              1013921856
```

I also tried `real_mnt->mnt_mountpoint->d_name.name` which doesn't seem to get back anything (output in the `MOUNTPOINT` field above):

```c
    // find mountpoint
    struct vfsmount *vmnt;
    bpf_probe_read_kernel(&vmnt, sizeof(vmnt), p->mnt);
    struct mount *real_mnt;
    bpf_probe_read_kernel(&real_mnt, sizeof(real_mnt), containerof(vmnt, struct mount, mnt));
    struct dentry *m_mp;
    bpf_probe_read_kernel(&m_mp, sizeof(m_mp), &real_mnt->mnt_mountpoint);
    struct qstr m_dname;
    bpf_probe_read_kernel(&m_dname, sizeof(m_dname), &m_mp->d_name);
    bpf_probe_read_kernel(&info.mountpoint, sizeof(info.mountpoint), m_dname.name);
```

TL;DR: I think I definitely did something wrong when trying to retrieve either `mnt_id` or `mnt_mountpoint->d_name.name`, but I cannot figure it out. Any pointers would be appreciated here!

**Comment by yonghong-song** [comment]
One possibility is kernel hit a fault when reading `m_dname.name`. Could you check the return value for
```
bpf_probe_read_kernel(&info.mountpoint, sizeof(info.mountpoint), m_dname.name)
```
?

**Comment by yzgyyang** [comment]
Thanks @yonghong-song - After checking the return values, I found out that the mistake is that I left out `&` before `p->mnt` in line 2 for `bpf_probe_read_kernel()`. The pointer chasing works after that. Can't believe it is simple as that! 🤦 

```c
    // find mountpoint
    struct vfsmount *vmnt;
    bpf_probe_read_kernel(&vmnt, sizeof(vmnt), &p->mnt);
    struct mount *real_mnt = containerof(vmnt, struct mount, mnt);
    bpf_probe_read_kernel(&info.mnt_id, sizeof(info.mnt_id), &real_mnt->mnt_id);
```
